kinto/core/cache/postgresql/__init__.py

```diff
@@ -1,12 +1,36 @@
 from __future__ import absolute_import
+from functools import wraps
 
 import os
+import time
 
 from kinto.core import logger
 from kinto.core.cache import CacheBase
 from kinto.core.storage.postgresql.client import create_from_config
+from kinto.core.storage.exceptions import BackendError
 from kinto.core.utils import json
 
+DELAY_BETWEEN_RETRIES_IN_SECONDS = 0.05
+MAX_RETRIES = 5
+
+
+def retry_on_failure(func):
+    import psycopg2
+
+    @wraps(func)
+    def wraps_func(self, *args, **kwargs):
+        tries = kwargs.pop('tries', 0)
+        try:
+            return func(self, *args, **kwargs)
+        except psycopg2.IntegrityError as e:
+            if tries < MAX_RETRIES:
+                delay = (tries - 1) * DELAY_BETWEEN_RETRIES_IN_SECONDS
+                if delay > 0:
+                    time.sleep(delay)
+                return wraps_func(self, tries=(tries + 1), *args, **kwargs)
+            raise BackendError(original=e)
```

Comment:
Actually the backends are not supposed to raise something else than `BackendError` because we want the user of the cache backend to be able to catch all exceptions regardless of the backend. I think we might have another bug to file about handling this for the cache backend.